### PR TITLE
Unrevert "Merge pull request #14704 from danhhz/persist_empty_antichain" 

### PR DESCRIPTION
**Pull request by danhhz**
We were seeing panics in nightly. See the second commit for details on the fix.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by danhhz** [comment]
Nightly run https://buildkite.com/materialize/nightlies/builds/1533

File: src/persist-client/src/read.rs [hunk 1]
```diff
@@ -752,7 +752,14 @@ where
                 .machine
                 .heartbeat_leased_reader(&self.reader_id, heartbeat_ts)
                 .await;
-            if !existed {
+            if !existed && !self.machine.state().collections.is_tombstone() {
+                // It's probably surprising to the caller that the shard
+                // becoming a tombstone expired this writer. Possibly the right
```
Comment (danhhz) [inline src/persist-client/src/read.rs:757]: fixed this "writer" -> "reader" and the "WriterId" -> "LeasedReaderId" below, but don't want to interrupt my CI runs, so leaving it on my local branch for now

**Review by pH14 — APPROVED** [review]
Fix for the expired handles here LGTM

**Comment by danhhz** [comment]
TFTR!

realized it was super straighforward to add a regression unit test, so added one (and verified that it fails before the fix commit)

**Pull request merged**